### PR TITLE
performance: cache system classpath in ClasspathScanner

### PR DESCRIPTION
**Pull request by cdietrich**
Signed-off-by: Christian Dietrich <christian.dietrich@itemis.de>

### CODE REVIEW

File: org.eclipse.xtext.common.types/src/org/eclipse/xtext/common/types/descriptions/ClasspathScanner.xtend [hunk 1]
```diff
@@ -52,16 +52,22 @@ class ClasspathScanner {
 		return ClassLoader.systemClassLoader.getDescriptors(true, packagePrefixes)
 	}
 	
+	var String[] systemClasspath = null
```
Comment (kthoms) [inline org.eclipse.xtext.common.types/src/org/eclipse/xtext/common/types/descriptions/ClasspathScanner.xtend]: Could this variable be accessed by different threads? Is synchronization required?

Comment (kthoms) [reply org.eclipse.xtext.common.types/src/org/eclipse/xtext/common/types/descriptions/ClasspathScanner.xtend]: Maybe just a theoretical issue. When 2 threads have null for that field, both would do the same computation and last wins. So if you think that does not matter then I'm fine with that.

Comment (szarnekow) [reply org.eclipse.xtext.common.types/src/org/eclipse/xtext/common/types/descriptions/ClasspathScanner.xtend]: What bugs me more is that we cache an array which is by definition not immutable. Callers may change the entries.

Comment (cdietrich) [reply org.eclipse.xtext.common.types/src/org/eclipse/xtext/common/types/descriptions/ClasspathScanner.xtend]: the array is in the signature

File: org.eclipse.xtext.common.types/src/org/eclipse/xtext/common/types/descriptions/ClasspathScanner.xtend [hunk 2]
```diff
@@ -52,16 +52,22 @@ class ClasspathScanner {
...
+	
 	def String[] getSystemClasspath() {
+		if (systemClasspath !== null) {
+			return systemClasspath;
```
Comment (szarnekow) [inline org.eclipse.xtext.common.types/src/org/eclipse/xtext/common/types/descriptions/ClasspathScanner.xtend]: `return systemClasspath.clone()` ?

File: org.eclipse.xtext.common.types/src/org/eclipse/xtext/common/types/descriptions/ClasspathScanner.xtend [hunk 3]
```diff
@@ -52,16 +52,22 @@ class ClasspathScanner {
...
+		}
 		try (val ScanResult scanResult = new ClassGraph()
 				.enableSystemJarsAndModules()
 				.addClassLoader(ClassLoader.getSystemClassLoader())
 				.scan()) {
 			val classpathURIs = scanResult.classpathURIs
-			return classpathURIs.map[path].filterNull
+			systemClasspath = classpathURIs.map[path].filterNull
+			return systemClasspath;
```
Comment (szarnekow) [inline org.eclipse.xtext.common.types/src/org/eclipse/xtext/common/types/descriptions/ClasspathScanner.xtend]: Also here: clone may prevent harm.

**Pull request merged**